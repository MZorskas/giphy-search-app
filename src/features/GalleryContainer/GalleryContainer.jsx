import React, { useRef, useState, useEffect } from 'react';
import './GalleryContainer.scss';
import ImageCard from 'components/ImageCard';
import ErrorMessage from 'components/ErrorMessage';
import PropTypes from 'prop-types';
import { elementsPerPage } from 'apis/giphy';

const GalleryContainer = ({ data, title, totalCount, fetchData }) => {
  const [elementsOffset, setElementsOffset] = useState(0);
  const [element, setElement] = useState(null);
  const [hasMore, setHasMore] = useState(true);

  const observer = useRef(
    new IntersectionObserver(
      (entries) => {
        const first = entries[0];
        if (first.isIntersecting)
          setElementsOffset((previousValue) => previousValue + elementsPerPage);
      },
      { threshold: 0.1 }
    )
  );

  useEffect(() => {
    const currentElement = element;
    const currentObserver = observer.current;
    if (currentElement) {
      currentObserver.observe(currentElement);
    }
    return () => {
      if (currentElement) {
        currentObserver.unobserve(currentElement);
      }
    };
  }, [element]);

  useEffect(() => {
    if (data) setHasMore(data.length < totalCount);
  }, [data, totalCount]);

  useEffect(() => {
    if (hasMore && elementsOffset !== 0) {
      // console.log('Should fetch now');
      fetchData(title, elementsOffset);
    }
  }, [elementsOffset, hasMore, fetchData, title]);

  return (
    <section className="gallery-container" data-testid="gallery-container">
      <div className="gallery-container__info-wrapper">
        <h2 className="gallery-container__title">
          {title}
          <span className="gallery-container__amount">{totalCount} GIFs</span>
        </h2>
      </div>

      {data?.length > 0 && (
        <div className="gallery-container__grid">
          {data?.map((el, index) => {
            if (data.length === index + 1) {
              return (
                <ImageCard
                  elementRef={setElement}
                  key={index}
                  images={el.images}
                  alt={el.title}
                />
              );
            } else {
              return (
                <ImageCard key={index} images={el.images} alt={el.title} />
              );
            }
          })}
        </div>
      )}

      {totalCount === 0 && (
        <ErrorMessage
          className="gallery-container__error-message"
          message={`No GIFs found for ${title}`}
        />
      )}
    </section>
  );
};

GalleryContainer.propTypes = {
  title: PropTypes.string.isRequired,
  totalCount: PropTypes.number.isRequired,
  fetchData: PropTypes.func.isRequired,
  data: PropTypes.array.isRequired,
};

export default GalleryContainer;
