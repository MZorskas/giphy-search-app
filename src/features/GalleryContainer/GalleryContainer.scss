@import 'src/styles/mixins/index';
@import 'src/styles/base/base';

.gallery-container {
  display: flex;
  flex-direction: column;

  &__info-wrapper {
    display: flex;
    color: $color-font-primary;
    margin: 1.5rem 0;
  }

  &__title {
    @include text-style-h2;

    margin: 0;
    padding: 0;
  }

  &__amount {
    @include text-style-body-small;

    padding: 0 0 0 0.5rem;
    color: $color-font-secondary;
  }

  &__grid {
    display: grid;
    grid-gap: 4rem;
    grid-template-columns: repeat(auto-fill, minmax(25rem, 1fr));
  }

  &__loader {
    color: $color-primary;
  }
}
