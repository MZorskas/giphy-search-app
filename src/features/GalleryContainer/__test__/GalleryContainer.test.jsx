import React from 'react';
import GalleryContainer from '../GalleryContainer';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const intersectionObserverMock = () => ({
  observe: () => null,
});

window.IntersectionObserver = jest
  .fn()
  .mockImplementation(intersectionObserverMock);

jest.mock('components/ImageCard', () => (props) => (
  <mock-ImageCard data-testId="mock-image-card" {...props} />
));

const mockfetchData = jest.fn();
const totalCount = 10;
const title = 'Title';
const errorMessage = `No GIFs found for ${title}`;
const altText = 'Test';
const data = [
  {
    images: { url: 'url1' },
    title: altText,
  },
  {
    images: { url: 'url2' },
    title: altText,
  },
  {
    images: { url: 'url3' },
    title: altText,
  },
];

describe('GalleryContainer component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');
    expect(screen.queryByTestId('gallery-container')).not.toBeInTheDocument();
    render(<GalleryContainer data={[]} />);
    expect(screen.queryByTestId('gallery-container')).toBeInTheDocument();
    expect(spy).toHaveBeenCalled();
  });

  it('renders without propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');
    expect(screen.queryByTestId('gallery-container')).not.toBeInTheDocument();
    render(
      <GalleryContainer
        data={[]}
        fetchData={mockfetchData}
        title={title}
        totalCount={0}
      />
    );
    expect(screen.queryByTestId('gallery-container')).toBeInTheDocument();
    expect(spy).not.toHaveBeenCalled();
  });

  it('displays title and totalCount correctly', () => {
    expect(screen.queryByText(title)).not.toBeInTheDocument();
    expect(screen.queryByText(`${totalCount} GIFs`)).not.toBeInTheDocument();
    render(
      <GalleryContainer
        data={[]}
        fetchData={mockfetchData}
        title={title}
        totalCount={totalCount}
      />
    );
    expect(screen.queryByText(title)).toBeInTheDocument();
    expect(screen.queryByText(`${totalCount} GIFs`)).toBeInTheDocument();
  });

  it('renders ImageCard elements', () => {
    expect(screen.queryByTestId('mock-image-card')).not.toBeInTheDocument();
    render(
      <GalleryContainer
        data={data}
        fetchData={mockfetchData}
        title={title}
        totalCount={totalCount}
      />
    );
    expect(screen.getAllByTestId('mock-image-card')[0]).toHaveAttribute(
      'alt',
      data[0].title
    );
    expect(screen.getAllByTestId('mock-image-card').length).toBe(data.length);
  });

  it('displays ErrorMessage if totalCount is 0', () => {
    expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
    const { rerender } = render(
      <GalleryContainer
        data={[]}
        fetchData={mockfetchData}
        title={title}
        totalCount={totalCount}
      />
    );
    expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
    rerender(
      <GalleryContainer
        data={[]}
        fetchData={mockfetchData}
        title={title}
        totalCount={0}
      />
    );
    expect(screen.queryByText(errorMessage)).toBeInTheDocument();
  });

  it('displays ErrorMessage with correct title', () => {
    expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();

    const { rerender } = render(
      <GalleryContainer
        data={[]}
        fetchData={mockfetchData}
        title={title}
        totalCount={0}
      />
    );

    expect(
      screen.queryByText(`No GIFs found for ${title}`)
    ).toBeInTheDocument();

    rerender(
      <GalleryContainer
        data={[]}
        fetchData={mockfetchData}
        title="Test"
        totalCount={0}
      />
    );
    expect(
      screen.queryByText(`No GIFs found for ${title}`)
    ).not.toBeInTheDocument();

    expect(screen.queryByText(`No GIFs found for Test`)).toBeInTheDocument();
    expect;
  });

  it('matches snapshot without data', () => {
    const tree = renderer
      .create(
        <GalleryContainer
          data={[]}
          fetchData={mockfetchData}
          title="Test"
          totalCount={totalCount}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with data', () => {
    const tree = renderer
      .create(
        <GalleryContainer
          data={data}
          fetchData={mockfetchData}
          title="Test"
          totalCount={totalCount}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with ErrorMessage', () => {
    const tree = renderer
      .create(
        <GalleryContainer
          data={[]}
          fetchData={mockfetchData}
          title="Test"
          totalCount={0}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
