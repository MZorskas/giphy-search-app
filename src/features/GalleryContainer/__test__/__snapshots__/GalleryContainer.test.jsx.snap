// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryContainer component matches snapshot with ErrorMessage 1`] = `
<section
  className="gallery-container"
  data-testid="gallery-container"
>
  <div
    className="gallery-container__info-wrapper"
  >
    <h2
      className="gallery-container__title"
    >
      Test
      <span
        className="gallery-container__amount"
      >
        0
         GIFs
      </span>
    </h2>
  </div>
  <p
    className="error-message gallery-container__error-message"
    data-testid="error-message"
  >
    No GIFs found for Test
  </p>
</section>
`;

exports[`GalleryContainer component matches snapshot with data 1`] = `
<section
  className="gallery-container"
  data-testid="gallery-container"
>
  <div
    className="gallery-container__info-wrapper"
  >
    <h2
      className="gallery-container__title"
    >
      Test
      <span
        className="gallery-container__amount"
      >
        10
         GIFs
      </span>
    </h2>
  </div>
  <div
    className="gallery-container__grid"
  >
    <mock-ImageCard
      alt="Test"
      data-testId="mock-image-card"
      images={
        Object {
          "url": "url1",
        }
      }
    />
    <mock-ImageCard
      alt="Test"
      data-testId="mock-image-card"
      images={
        Object {
          "url": "url2",
        }
      }
    />
    <mock-ImageCard
      alt="Test"
      data-testId="mock-image-card"
      elementRef={[Function]}
      images={
        Object {
          "url": "url3",
        }
      }
    />
  </div>
</section>
`;

exports[`GalleryContainer component matches snapshot without data 1`] = `
<section
  className="gallery-container"
  data-testid="gallery-container"
>
  <div
    className="gallery-container__info-wrapper"
  >
    <h2
      className="gallery-container__title"
    >
      Test
      <span
        className="gallery-container__amount"
      >
        10
         GIFs
      </span>
    </h2>
  </div>
</section>
`;
